ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs

```diff
@@ -0,0 +1,482 @@
+                     PBftParams (..))
+import           Ouroboros.Consensus.Protocol.Praos (PraosMockCrypto,
+                     PraosParams (..))
+import           Ouroboros.Consensus.Util.Orphans ()
+
+import qualified Test.Dynamic.Ref.RealPBFT as Ref.RealPBFT
+import           Test.Dynamic.Util (consensusExpected)
+import           Test.Dynamic.Util.NodeJoinPlan
+import           Test.Dynamic.Util.NodeTopology
+
+import           Test.Util.Orphans.Arbitrary ()
+
+{-------------------------------------------------------------------------------
+  Configuring tests
+-------------------------------------------------------------------------------}
+
+data TestConfig blk = TestConfig
+  { adhocTestConfig :: !(AdhocTestConfig blk)
+  , nodeJoinPlan    :: !NodeJoinPlan
+  , nodeTopology    :: !NodeTopology
+  , numCoreNodes    :: !NumCoreNodes
+  , numSlots        :: !NumSlots
+  }
+
+deriving instance TestableBlock blk => Show (TestConfig blk)
+
+instance TestableBlock blk => Arbitrary (TestConfig blk) where
+    arbitrary = genTestConfig
+    shrink    = shrinkTestConfigFreely
+
+-- | Protocol-dependent configuration
+--
+class Show (AdhocTestConfig blk) => TestableBlock blk where
+    type AdhocTestConfig blk
+
+    genTestConfig    :: Gen (TestConfig blk)
+    shrinkTestConfig :: TestConfig blk -> [TestConfig blk]
+
+    -- | List any inconsistencies between 'adhocTestConfig' and the other
+    -- fields
+    --
+    -- EG 'adhocTestConfig' might include on 'NumCoreNodes'
+    --
+    adhocInconsistenciesTestConfig :: TestConfig blk -> [String]
+
+defaultGenTestConfig ::
+    AdhocTestConfig blk -> NumCoreNodes -> NumSlots -> Gen (TestConfig blk)
+defaultGenTestConfig adhocTestConfig numCoreNodes numSlots = do
+    nodeJoinPlan <- genNodeJoinPlan numCoreNodes numSlots
+    nodeTopology <- genNodeTopology numCoreNodes
+    pure TestConfig
+      { adhocTestConfig
+      , nodeJoinPlan
+      , nodeTopology
+      , numCoreNodes
+      , numSlots
+      }
+
+-- | Shrink, including the number of nodes and slots
+--
+shrinkTestConfigFreely :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigFreely TestConfig
+  { adhocTestConfig
+  , numCoreNodes
+  , numSlots
+  , nodeJoinPlan
+  , nodeTopology
+  } =
+    dropId $
+    [ TestConfig
+        { adhocTestConfig
+        , nodeJoinPlan    = p'
+        , nodeTopology    = top'
+        , numCoreNodes    = n'
+        , numSlots        = t'
+        }
+    | n'             <- andId shrink numCoreNodes
+    , t'             <- andId shrink numSlots
+    , let truncateP   = truncateNodeJoinPlan nodeJoinPlan n' (numSlots, t')
+    , let truncateTop = truncateNodeTopology nodeTopology n'
+    , p'             <- andId shrinkNodeJoinPlan truncateP
+    , top'           <- andId shrinkNodeTopology truncateTop
+    ]
+
+-- | Shrink, including the number of slots but not number of nodes
+--
+shrinkTestConfigSlotsOnly :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigSlotsOnly TestConfig
```

Comment:
I simplified it so that we have two shrinks, currently:

  * One in the `Arbitrary TestConfig` instance that shrinks all fields.
  * One in `RealPBFT.hs` that shrinks everything except num core nodes, since doing that might violate the chain density invariant that the custom `NodeJoinPlan` generator in that module ensures.

`Praos.hs` and `LeaderSchedule.hs` use nested `forAllShrink`s, since they have some fixed parameters (which have always been fixed and I haven't yet considered varying).